core/virtualhostname/virtualhostname.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package virtualhostname
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v5"
+)
+
+const (
+	// ErrNoMatch indicates that the hostname could not be parsed.
+	ErrNoMatch = errors.ConstError("could not parse hostname")
+)
+
+// HostnameTarget defines what kind of infrastructure the user is targeting.
+type HostnameTarget int
+
+const (
+	// MachineTarget defines a machine as the target.
+	MachineTarget HostnameTarget = iota
+
+	// UnitTarget defines a unit (K8s or machine)
+	// as the target.
+	UnitTarget
+
+	// ContainerTarget defines a container within a
+	// unit as the target.
+	ContainerTarget
+)
+
+// Info returns a breakdown of a virtual
+// hostname into it's constituent parts.
+// The target field indicates what kind of
+// hostname was parsed which will indicate
+// that some fields are empty.
+type Info struct {
+	modelUUID string
+	machine   int
+	unit      int
+	app       string
+	container string
+	target    HostnameTarget
+}
+
+// Unit returns the unit name, appropriate
+// for use in state methods.
+func (i Info) Unit() string {
+	return fmt.Sprintf("%s/%d", i.app, i.unit)
+}
+
+// App returns the application name.
+func (i Info) Application() string {
+	return i.app
+}
+
+// Container returns the container name.
+func (i Info) Container() string {
```

Comment:
These getters only return a valid result if the target matches. ie it makes no sense to call `Container()` if target is `Machine`. We should look to return an error or at the least a bool to indicate if the value is valid.

Reply:
Added a bool to this function and others. The `ModelUUID` getter is appropriate for all target types so I left the return type as just `string`, but if we want a bool there too for consistency let me know.